airgen/src/lib.rs

```diff
@@ -302,4 +334,79 @@ impl<'a> ASMPILConverter<'a> {
             is_permutation,
         }
     }
+
+    // Process machine parameters.
+    // Allows machines passed as argument to be referenced.
+    // TODO: support some elementary pil types as a machine parameter?
+    fn handle_parameters(&mut self, params: MachineParams, values: &Vec<Expression>) {
+        if params.0.len() != values.len() {
+            panic!(
+                "wrong number of arguments for machine `{}`: got {} expected {}",
+                self.location,
+                values.len(),
+                params.0.len()
+            );
+        }
+        for (param, value) in params.0.iter().zip(values) {
+            let ty = AbsoluteSymbolPath::default().join(param.ty.clone().unwrap());
+            if let Some(Item::Machine(_)) = self.items.get(&ty) {
+                // param is a machine, we need to find the actual instance
```

Comment:
Wait but shouldn't the path canonicalizer have turned this into an absolute name, why do we need to search?

Reply:
The code here is looking for the proper `location` of a submachine parameter.
Say you have a `memory: Memory` parameter that as used to passe a value of `foo`.
We need to go up the machine tree (is `foo` a local submachine or another parameter in the parent?) until we find the actual submachine declaration.
Should the path canonicalizer do this?

Reply:
Ok, yeah, I understood that better in the meantime. It's not going up the module tree, it's going up the instantiation chain. It's just by convention, that the locations are hierarchical like that. I'll have to review this properly again.